core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java

```diff
@@ -178,15 +181,19 @@ private TableMetadata getTableMetadata(Scan scan) throws ExecutionException {
   }
 
   public Optional<List<Key>> get(Scan scan) {
-    if (isWriteSetOverlappedWith(scan)) {
-      throw new IllegalArgumentException("reading already written data is not allowed");
-    }
     if (scanSet.containsKey(scan)) {
       return Optional.ofNullable(scanSet.get(scan));
     }
     return Optional.empty();
   }
 
+  public void verify(Scan scan) {
+    if ((ScalarDbUtils.isRelational(scan) && isWriteSetOverlappedWithRelational(scan))
+        || (!ScalarDbUtils.isRelational(scan) && isWriteSetOverlappedWith(scan))) {
```

Comment:
Sorry, one more point. To avoid duplicate calculations, I think we should do the following:
```suggestion
    boolean isRelational = ScalarDbUtils.isRelational(scan);
    if ((isRelational && isWriteSetOverlappedWithRelational(scan))
        || (!isRelational) && isWriteSetOverlappedWith(scan))) {
```

Reply:
Right, I will fix it. Thank you!

Reply:
@brfrn169 Fixed in 0c90275. PTAL if you have time.